### PR TITLE
Fix: Integration test does not cover remote reindexing (#2789) 

### PR DESCRIPTION
**Pull request by jessebrennan**
https://github.com/DataBiosphere/azul/issues/2789

### Author

- [x] PR title references issue
- [x] PR title matches issue title (preceded by `Fix: ` for bugs)   <sub>or there is a good reason why they're different</sub>
- [x] Title of main commit references issue
- [x] PR is connected to Zenhub issue and description links to issue

### Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

### Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] Added announcement to PR description                  <sub>or this PR does not require announcement</sub>
- [x] Added checklist items for additional operator tasks   <sub>or this PR does not require additional tasks</sub>

### Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR does not touch requirements*.txt</sub>

### Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

### Primary reviewer (after approval)

- [x] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [x] Pushed merge commit to Github

### Operator (after pushing the merge commit)

- [ ] Made announcement requested by author                 <sub>or PR description does not contain an announcement</sub>
- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate       <sub>or or this PR is authored by lead</sub>
- [ ] Pushed merge commit to Gitlab                         <sub>or or merge commit can be pushed later, with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Build passes on Gitlab

### Operator (reindex) 

- [ ] Started reindex in target deployment                  <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures             <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in target deployment              <sub>or this PR does not require reindexing</sub>

### Operator

- [ ] Unassigned PR

### CODE REVIEW

File: test/integration_test.py [hunk 1]
```diff
@@ -214,36 +210,98 @@ def managed_access_sources_by_catalog(self) -> Dict[CatalogName,
...
+                fqids = self.azul_client.filter_obsolete_bundle_versions(fqids)
+                # By the pigeonhole principle, this condition is guaranteed to
+                # be true eventually, and likely with few iterations
+                if len(fqids) > 0:
+                    return prefix, fqids
+                else:
+                    # Increment to the next prefix without carrying over
+                    prefix = hex(int(prefix, 16) + 1)[-prefix_len:]
+
+    def _choose_prefix_2(self,
```
Comment (jessebrennan) [inline test/integration_test.py:276]: This is an earlier attempt which I don't like as much and will likely remove after Hannes reviews, assuming he prefers the other implementation.

**Comment by noah-aviel-dove** [comment]
Since I'm presumably taking over this PR with Jesse leaving, I'm going to author a `fixup!` commit instead of leaving a review, then ask for peer review from someone else.

**Review by noah-aviel-dove — COMMENTED** [review]
There are things I thought could be improved, but I think my review is invalided because I'm taking over as the author of this PR. I think the next reviewer should look at both Jesse's original commit and my proposed `fixup!` and see whether they think I managed to improve on it.

**Comment by noah-aviel-dove** [comment]
Note to self: do not squash fixups, keep my and Jesse's changes separate as a split commit.

**Review by amarjandu — CHANGES_REQUESTED** [review]
I had a failure when running these IT changes. Can you double check that it runs on your machine?

Also I think this comment is applicable here too. https://github.com/DataBiosphere/azul/pull/3475#pullrequestreview-779288367

```
======================================================================
FAIL: test_can_bundle_configured_catalogs (integration_test.CanBundleScriptIntegrationTest) (catalog='it2', repository=Config.Catalog.Plugin(name='tdr'))
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/amar/dev/hca/azul/test/integration_test.py", line 1301, in test_can_bundle_configured_catalogs
    self._test_catalog(catalog)
  File "/Users/amar/dev/hca/azul/test/integration_test.py", line 1259, in _test_catalog
    fqid = self.bundle_fqid(catalog.name)
  File "/Users/amar/dev/hca/azul/test/integration_test.py", line 1323, in bundle_fqid
    source, prefix, bundle_fqids = next(self._list_bundles(catalog, min_bundles=1))
  File "/Users/amar/dev/hca/azul/test/integration_test.py", line 226, in _list_bundles
    prefix, new_fqids = self._choose_prefix(catalog, source)
  File "/Users/amar/dev/hca/azul/test/integration_test.py", line 276, in _choose_prefix
    assert fqids, fqids
AssertionError: []
```

File: src/azul/azulclient.py [hunk 2]
```diff
@@ -277,6 +278,14 @@ def remote_reindex_partition(self, message: JSON) -> None:
         logger.info('Successfully queued %i notification(s) for prefix %s of '
                     'source %r', num_messages, prefix, source)
 
+    def batch_queue_messages(self, messages: JSONs):
```
[the remark this answers is not in the document]
Comment (noah-aviel-dove) [reply src/azul/azulclient.py]: I was eventually able to reproduce that `AssertionError` by setting the random seed of `0`.

[the remark this answers is not in the document]
Comment (noah-aviel-dove) [reply src/azul/azulclient.py]: Well, now I can't reproduce it anymore but I did see it so I believe it's real even if I don't understand why it's inconsistent. I suspect, but cannot confirm, that this is due to https://github.com/DataBiosphere/azul/issues/3561, which I've already marked as a blocker for this ticket.

[the remark this answers is not in the document]
Comment (noah-aviel-dove) [reply src/azul/azulclient.py]: Yeah, I keep trying and can't reproduce this. If it happens again we need to record the full log output.

[the remark this answers is not in the document]
Comment (amarjandu) [reply src/azul/azulclient.py]: I was not able to replicate the failure when I re-ran the IT. :(

File: test/integration_test.py [hunk 3]
```diff
@@ -769,14 +772,24 @@ def _prepare_notifications(self, catalog: CatalogName) -> Tuple[JSONs,
...
             reindex_notifications.append({
                 'action': 'reindex',
                 'catalog': catalog,
                 'source': source.to_json(),
                 'prefix': prefix
             })
             fqids.update(new_fqids)
+
+        for source_bundle_info in self._list_managed_access_bundles(catalog):
+            update(*source_bundle_info)
```
Comment (amarjandu) [inline test/integration_test.py:786]: the use of `*source_bundle_info` hides a type warning that the returned tuple is a `(SourceRef, str, List)` per the type hint of the `_list_*` methods. while the `update` method expects the `fqid` arg to be a `set`

Screen Shot 2021-11-04 at 1 07 10 PM

**Review by amarjandu — CHANGES_REQUESTED** [review]

File: test/integration_test.py (hunk 3 above)
Comment (noah-aviel-dove) [reply test/integration_test.py]: The type annotation was too strict anyway, I incorrectly assumed `set.update` required a `set` but it just needs an `Iterable`. Nice catch.

**Review by amarjandu — APPROVED** [review]
Spoke with Noah about removing the skip for duplicate notifications, opted to keep the PR as is for the time being.

File: test/integration_test.py [hunk 4]
```diff
@@ -214,36 +211,101 @@ def managed_access_sources_by_catalog(self) -> Dict[CatalogName,
...
+            prefix = bundle_fqid.uuid[len(common):8]
+            assert prefix != '', prefix
+            new_fqids = self.azul_client.list_bundles(catalog, source, prefix)
+            yield source, prefix, new_fqids
+
+    def _choose_prefix(self,
+                       catalog: CatalogName,
+                       source: SourceRef
+                       ) -> Tuple[str, List[SourcedBundleFQID]]:
+        fqids = self.azul_client.list_bundles(catalog, source, '')
```
Comment (hannes-ucsc) [inline test/integration_test.py:255]: This looks like it lists all bundles in the source. If that's true, this approach would be infeasible, as I mentioned in a previous parking lot discussion with you. If we really need to count bundles, we need to add a plugin method for just that.

But: I remember giving different directives to the previous author. We currently (on develop) hard-code the prefix and the partition prefix length in the configuration (with the help of a heuristic) to get a reasonable number of bundles per partition. We decided to add a test to check the hardcoded prefix to make sure it doesn't produce an empty source. We should extend that check to ensure it doesn't produce overly small or overly large partitions. Once we have that, we simply pick a random partition from a set of randomly selected sources and we're done.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Comment by noah-aviel-dove** [comment]
I had written a `fixup` but it was too difficult to keep it intact while rebasing, since the order of the chain with https://github.com/DataBiosphere/azul/pull/3612 was reversed since the last review.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
What's the purpose of the split commit? Both commits have same author and there is not explanation in the commit message body motivating the split.

https://github.com/DataBiosphere/azul/blob/develop/CONTRIBUTING.rst#33split-commits

**Comment by noah-aviel-dove** [comment]
First commit was authored by @jessebrennan. I had to repeatedly resolve conflicts in the 1st commit so I guess GitHub made me the author. I'm not sure whether to just squash or try to restore Jesse's authorship.

File: test/integration_test.py [hunk 5]
```diff
@@ -300,25 +332,25 @@ def test_indexing(self):
...
-                num_duplicates = self.num_bundles // 2
-                notifications = list(self.notifications.values())
-                # Index some bundles again to test that we handle duplicate additions.
-                # Note: random.choices() may pick the same element multiple times so
-                # some notifications will end up being sent three or more times.
-                notifications.extend(self.random.choices(notifications, k=num_duplicates))
-                return notifications
+                # FIXME: Add back duplicates
+                #        https://github.com/DataBiosphere/azul/issues/3558
+                if False:
```
Comment (hannes-ucsc) [inline test/integration_test.py:346]: Neither the ticket nor the comment explains why this has to be disabled.

File: src/azul/azulclient.py [hunk 6]
```diff
@@ -398,6 +402,9 @@ def delete_bundle(self, catalog: CatalogName, bundle_uuid, bundle_version):
     def deindex(self, catalog: CatalogName, sources: Iterable[str]):
         plugin = self.repository_plugin(catalog)
         sources = [plugin.resolve_source(s) for s in sources]
+        self.deindex_source_refs(catalog, sources)
+
+    def deindex_source_refs(self, catalog: CatalogName, sources: Iterable[SourceRef]):
```
Comment (hannes-ucsc) [inline src/azul/azulclient.py:407]: Only one call site, so no need to extract.

File: test/integration_test.py [hunk 7]
```diff
@@ -1287,8 +1312,8 @@ def test_can_bundle_canned_repository(self):
             self._test_catalog(mock_catalog)
 
     def bundle_fqid(self, catalog: CatalogName) -> SourcedBundleFQID:
-        bundle_fqids = self._list_bundles(catalog, max_bundles=1)
-        return one(bundle_fqids)
+        source, prefix, bundle_fqids = next(self._list_bundles(catalog, min_bundles=1))
```
Comment (hannes-ucsc) [inline test/integration_test.py:1315]: Is it a problem that this exhausts all sources?

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
```diff
Index: src/azul/azulclient.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/azulclient.py b/src/azul/azulclient.py
--- a/src/azul/azulclient.py	(revision 032627b115d3969241480984bdf0eda5c04fdedd)
+++ b/src/azul/azulclient.py	(date 1641518164668)
@@ -266,11 +266,11 @@
             }
             for bundle_fqid in bundle_fqids
         )
-        num_messages = self.batch_queue_messages(messages)
+        num_messages = self.queue_notifications(messages)
         logger.info('Successfully queued %i notification(s) for prefix %s of '
                     'source %r', num_messages, prefix, source)
 
-    def batch_queue_messages(self, messages: Iterable[JSON]) -> int:
+    def queue_notifications(self, messages: Iterable[JSON]) -> int:
         num_messages = 0
         for batch in chunked(messages, 10):
             entries = [
Index: test/integration_test.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/integration_test.py b/test/integration_test.py
--- a/test/integration_test.py	(revision 032627b115d3969241480984bdf0eda5c04fdedd)
+++ b/test/integration_test.py	(date 1641519379194)
@@ -13,6 +13,9 @@
     BytesIO,
     TextIOWrapper,
 )
+from itertools import (
+    starmap,
+)
 import json
 import logging
 import os
@@ -252,9 +255,7 @@
             if total_bundles < min_bundles:
                 total_bundles += len(new_fqids)
                 yield source, prefix, new_fqids
-            else:
-                # Keep looping to verify the size of a partition for all sources
-                continue
+            # Keep looping to verify the size of a partition for all sources
         if total_bundles < min_bundles:
             log.warning('Checked all sources and found only %d bundles instead of the '
                         'expected minimum %d', total_bundles, min_bundles)
@@ -376,7 +377,7 @@
 
         if index:
             for catalog in catalogs:
-                self.azul_client.batch_queue_messages(catalog.notifications)
+                self.azul_client.queue_notifications(catalog.notifications)
             _wait_for_indexer()
             for catalog in catalogs:
                 self._assert_catalog_complete(catalog=catalog.name,
@@ -746,30 +747,26 @@
 
     def _prepare_notifications(self, catalog: CatalogName) -> Tuple[JSONs,
                                                                     Set[SourcedBundleFQID]]:
-        reindex_notifications = []
-        fqids = set()
+        bundle_fqids = set()
+        notifications = []
 
         def update(source: SourceRef,
                    prefix: str,
-                   new_fqids: Iterable[SourcedBundleFQID]):
-            reindex_notifications.append({
+                   partition_bundle_fqids: Iterable[SourcedBundleFQID]):
+            bundle_fqids.update(partition_bundle_fqids)
+            notifications.append({
                 'action': 'reindex',
                 'catalog': catalog,
                 'source': source.to_json(),
                 'prefix': prefix
             })
-            fqids.update(new_fqids)
 
-        for ref, prefix, new_fqids in self._list_managed_access_bundles(catalog):
-            update(ref, prefix, new_fqids)
-        num_bundles = self.min_bundles - len(fqids)
+        starmap(update, self._list_managed_access_bundles(catalog))
+        num_bundles = self.min_bundles - len(bundle_fqids)
         log.info('Selected %d bundles to satisfy managed access coverage; '
-                 'selecting at least %d more', len(fqids), num_bundles)
-        for ref, prefix, new_fqids in self._list_bundles(catalog,
-                                                         min_bundles=num_bundles):
-            update(ref, prefix, new_fqids)
-
-        return reindex_notifications, fqids
+                 'selecting at least %d more', len(bundle_fqids), num_bundles)
+        starmap(update, self._list_bundles(catalog, min_bundles=num_bundles))
+        return notifications, bundle_fqids
 
     def _assert_catalog_complete(self,
                                  catalog: CatalogName,
```

File: test/integration_test.py (hunk 7 above)
Comment (hannes-ucsc) [reply test/integration_test.py:1315]: Ah, maybe it doesn't.

**Comment by noah-aviel-dove** [comment]
The current restraints on partition size are impossible to meet for some snapshots. For example `datarepo-dev-4de1b9fd.hca_dev_16ed4ad8731946b288596fe1c1d73a82__20210830_20210903` has 28 subgraphs, exceeding the upper limit of 24 (16*1.5). Adding any 1-character prefix reduces the number of subgraphs to at most 4, violating the lower bound of 8 (16/2).

File: test/integration_test.py (hunk 5 above)
Comment (noah-aviel-dove) [reply test/integration_test.py]: I'm not sure myself. Added back duplicates and it seems to work fine but I'm worried I'm missing something.

File: test/integration_test.py [hunk 8]
```diff
@@ -235,7 +235,7 @@ def _list_partition_bundles(self,
         else:
             expected_partition_size = 16
         lower = expected_partition_size / 2
-        upper = expected_partition_size * 1.5
+        upper = expected_partition_size * 2
```
Comment (noah-aviel-dove) [inline test/integration_test.py:238]: Smallest number that worked without going to an unreasonable number of decimal places.

File: test/integration_test.py (hunk 5 above)
Comment (jessebrennan) [reply test/integration_test.py]: As I recall, @hannes-ucsc had explicitly asked me to disable this, and keep things as simple as possible for a mvp.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
```diff
Index: test/integration_test.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/integration_test.py b/test/integration_test.py
--- a/test/integration_test.py	(revision e7fe310bf7afec8da740fb240aaffa23b8095422)
+++ b/test/integration_test.py	(date 1642620747312)
@@ -226,9 +226,9 @@
         """
         source = self.azul_client.repository_plugin(catalog).resolve_source(source)
         partition_prefixes = list(source.spec.prefix.effective.partition_prefixes())
-        prefix = self.random.choice(partition_prefixes)
-        fqids = self.azul_client.list_bundles(catalog, source, prefix)
-        partition = f'Partition {prefix!r} of source {source.spec!r}'
+        partition_prefix = self.random.choice(partition_prefixes)
+        fqids = self.azul_client.list_bundles(catalog, source, partition_prefix)
+        partition = f'Partition {partition_prefix!r} of source {source.spec!r}'
         self.assertGreater(len(fqids), 0, partition + ' is empty')
         if config.is_main_deployment(config.deployment_stage):
             expected_partition_size = 512
@@ -238,12 +238,14 @@
         upper = expected_partition_size * 2
         assert 0 < lower < upper, expected_partition_size
         self.assertLessEqual(len(fqids), upper, partition + ' is too large')
-        # Some snapshots have fewer bundles in total than the
-        # desired partition size
-        if prefix:
+        # REVIEW: not sure about this, though
+        if source.prefix and partition_prefix:
             self.assertGreaterEqual(lower, len(fqids), partition + ' is too small')
+        else:
+            log.warning('Cannot enforce minimum parition size since source '
+                        '%r has too few bundles in total.', source)
 
-        return source, prefix, fqids
+        return source, partition_prefix, fqids
 
     def _list_bundles(self,
                       catalog: CatalogName,
```

**Comment by noah-aviel-dove** [comment]
Preparation of notifications succeeds with `dev` selected.

File: test/integration_test.py [hunk 9]
```diff
@@ -216,41 +216,99 @@ def managed_access_sources_by_catalog(self) -> Dict[CatalogName,
...
+        source = plugin.resolve_source(source)
+        prefix = source.spec.prefix
+        partition_prefixes = list(prefix.effective.partition_prefixes())
+        partition_prefix = self.random.choice(partition_prefixes)
+        effective_prefix = source.spec.prefix.common + partition_prefix
+        fqids = self.azul_client.list_bundles(catalog, source, partition_prefix)
+        bundle_count = len(fqids)
+        partition = f'Partition {effective_prefix!r} of source {source.spec}'
+        if config.is_main_deployment(config.deployment_stage):
+            # For snapshots that use partitioning, 512 is the maximum allowed
```
Comment (hannes-ucsc) [inline test/integration_test.py:234]:
```suggestion
            # For sources that use partitioning, 512 is the maximum allowed
```

File: test/integration_test.py [hunk 10]
```diff
@@ -216,41 +216,99 @@ def managed_access_sources_by_catalog(self) -> Dict[CatalogName,
...
+            upper = 512
+            if effective_prefix:
+                lower = upper // 16
+                if len(fqids) < lower:
+                    # If bundle UUIDs were deterministically uniform, we could
+                    # directly assert a minimum partition size, but since
+                    # they're random, a given partition may fail to exceed this
+                    # minimum even with an optimal choice of partition prefix
+                    # length.
+                    log.warning('%s is too small', partition)
```
Comment (hannes-ucsc) [inline test/integration_test.py:245]:
```suggestion
                    log.warning('With %i bundles, %s is too small', len(fqids), partition)
```

File: test/integration_test.py [hunk 11]
```diff
@@ -216,41 +216,99 @@ def managed_access_sources_by_catalog(self) -> Dict[CatalogName,
...
+                        lower = 1
+                    else:
+                        # For plugins that support efficiently counting bundles
+                        # across all partitions simultaneously, we can check
+                        # whether the chosen partition is outlier by determining
+                        # the *average* partition size.
+                        bundle_count = sum(counts.values()) / len(counts)
+            else:
+                # Snapshots too small to be split into more than one partition
+                # may have as few as one bundle in total
```
Comment (hannes-ucsc) [inline test/integration_test.py:257]:
```suggestion
                # Sources too small to be split into more than one partition
                # may have as few as one bundle in total.
```

File: test/integration_test.py [hunk 12]
```diff
@@ -216,41 +216,99 @@ def managed_access_sources_by_catalog(self) -> Dict[CatalogName,
...
+                lower = 1
+        else:
+            # For deployments that don't use partitioning, the desired number of
```
Comment (hannes-ucsc) [inline test/integration_test.py:260]:
```suggestion
            # Personal deployments typically don't use partitioning and the desired number of
```

File: test/integration_test.py [hunk 13]
```diff
@@ -216,41 +216,99 @@ def managed_access_sources_by_catalog(self) -> Dict[CatalogName,
...
+        bundle_count = len(fqids)
+        partition = f'Partition {effective_prefix!r} of source {source.spec}'
+        if config.is_main_deployment(config.deployment_stage):
+            # For snapshots that use partitioning, 512 is the maximum allowed
+            # partition size.
+            upper = 512
+            if effective_prefix:
+                lower = upper // 16
+                if len(fqids) < lower:
+                    # If bundle UUIDs were deterministically uniform, we could
```
Comment (hannes-ucsc) [inline test/integration_test.py:240]:
```suggestion
                    # If bundle UUIDs were uniformly distributed by prefix, we could
```

File: test/integration_test.py [hunk 14]
```diff
@@ -216,41 +216,99 @@ def managed_access_sources_by_catalog(self) -> Dict[CatalogName,
...
+                    # directly assert a minimum partition size, but since
+                    # they're random, a given partition may fail to exceed this
```
Comment (hannes-ucsc) [inline test/integration_test.py:242]:
```suggestion
                    # they're not, a given partition may fail to exceed this
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Comment by achave11** [comment]
This PR requires non-trivial conflict resolution in `integration_test.py`. Re requesting author to resolve conflict.

File: src/azul/plugins/repository/tdr/__init__.py [hunk 15]
```diff
@@ -240,6 +240,23 @@ def list_bundles(self, source: TDRSourceRef, prefix: str) -> List[TDRBundleFQID]
...
+                        ) -> Dict[str, int]:
+        prefix = source.spec.prefix.effective
+        prefixes = [
+            prefix.common + partition_prefix
+            for partition_prefix in prefix.partition_prefixes()
+        ]
+        assert prefixes, prefix
+        rows = self._run_sql(f'''
+            SELECT prefix, COUNT(links_id) AS subgraph_count
+            FROM {backtick(self._full_table_name(source.spec, "links"))}
```
[the remark this answers is not in the document]
Comment (noah-aviel-dove) [reply src/azul/plugins/repository/tdr/__init__.py]: No other doublequoted string literals found.

**Review by hannes-ucsc — APPROVED** [review]
